### PR TITLE
Efficient linear entropy

### PR DESCRIPTION
**Pull request by Jianbing-D**
# Fused Linear Cross Entropy Implementation

This PR introduces an efficient kernel implementation for fused linear (matrix multiplication) and cross entropy operations using Triton. The implementation provides both forward and backward passes with memory and computational optimizations.

## Key Features

- Fuses matrix multiplication with cross entropy computation
- Implements both entropy and cross-entropy calculation in a single pass
- Provides memory-efficient backward pass with two implementation strategies:
  - Total Fused MN: Fuses d_logits & d_hidden & d_weight computation with no intermediate storage (requires FP32 as gradient dtype)
  - Total Separate: Uses intermediate d_logits storage with no special precision requirements

### Equivalent Torch Logic
```python
def run_torch_entropy(hidden: torch.Tensor,
                    weight: torch.Tensor,
                    labels: torch.Tensor) -> typing.List[torch.Tensor]:
    logits = torch.matmul(hidden.to(torch.float32), weight.to(torch.float32)) # [num_tokens, vocab_size]
    pd = torch.nn.functional.softmax(logits, dim=-1) # [num_tokens, vocab_size]
    entropy_a = torch.logsumexp(logits, dim=-1) # [num_tokens]
    entropy_b = torch.sum(pd * logits, dim=-1) # [num_tokens]
    entropy = entropy_a - entropy_b
    logprobs = torch.nn.functional.cross_entropy(logits, labels) # [1]
    return logprobs, entropy
```

### Forward Pass
- Calculates both cross-entropy loss and per-token entropy values
- Fuse multiple operations into fewer kernels to reduce peak memory pressure
- Implements numerically stable softmax

### Backward Pass
- Optimizes memory usage through careful buffer management
- Fuse backward pass into fewer kernels for reduced peak memory pressure and better compute efficiency

## Testing
The implementation includes comprehensive tests that:
- Verify numerical correctness against PyTorch reference implementation
- Compare memory usage between kernel and reference implementations
- Measure and compare forward/backward pass latencies

## Usage Example
```python
from transformer_engine.pytorch.linear_cross_entropy import linear_cross_entropy

hidden = torch.randn(num_tokens, hidden_size, dtype=torch.bfloat16, device="cuda")
weight = torch.randn(hidden_size, vocab_size, dtype=torch.bfloat16, device="cuda")
labels = torch.randint(0, vocab_size, (num_tokens,), device="cuda")

loss, entropy = linear_cross_entropy(hidden, weight, labels, reduction="mean")
```

### CODE REVIEW

**Comment by Jianbing-D** [comment]
Previous PR was not signed off by me. Has to force push signed-off commits, and open a new PR for DCO checking.
https://github.com/NVIDIA/TransformerEngine/pull/1533

**Comment by Jianbing-D** [comment]
## Tested with linting
image

## Tested with pytest
image

**Comment by nagarajankarthik** [comment]
When I try to run the tests, I obtain the following results:

```
tests/pytorch/test_linear_cross_entropy.py
[INFO]: Forward pass: Torch implementation average time: 90.45 ms
[INFO]: Backward pass: torch implementation average time: 192.86 ms
[INFO]: Forward pass: Kernel implementation average time: 517.10 ms
[INFO]: Backward pass: kernel implementation average time: 555.55 ms
.
[INFO]: Torch Forward pass peak memory: 26718.27 MB
[INFO]: Torch Backward pass peak memory: 57095.80 MB
.
[INFO]: Kernel Forward pass peak memory: 1376.98 MB
[INFO]: Kernel Backward pass peak memory: 2984.55 MB
.                                                                                                                                                                                                                [INFO]: Forward pass: Torch implementation average time: 52.30 ms
[INFO]: Backward pass: torch implementation average time: 139.77 ms
[INFO]: Forward pass: Kernel implementation average time: 520.16 ms
[INFO]: Backward pass: kernel implementation average time: 554.70 ms
.
[INFO]: Torch Forward pass peak memory: 19123.71 MB
[INFO]: Torch Backward pass peak memory: 26718.09 MB
.
[INFO]: Kernel Forward pass peak memory: 1369.42 MB
[INFO]: Kernel Backward pass peak memory: 2984.37 MB
```
The tests were run using the [Nvidia NeMo docker container version 25.07](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo?version=25.07.gpt_oss) . Please find attached the list of environment variables.
[collect_env.txt](https://github.com/user-attachments/files/22401195/collect_env.txt)



It seems that the kernel implementation time is significantly larger than the corresponding time for PyTorch, unlike the screenshot in the earlier message. @Jianbing-D , could you please clarify the hardware and environment used for your Pytest runs?

**Comment by Jianbing-D** [comment]
> When I try to run the tests, I obtain the following results:
> 
> ```
> tests/pytorch/test_linear_cross_entropy.py
> [INFO]: Forward pass: Torch implementation average time: 90.45 ms
> [INFO]: Backward pass: torch implementation average time: 192.86 ms
> [INFO]: Forward pass: Kernel implementation average time: 517.10 ms
> [INFO]: Backward pass: kernel implementation average time: 555.55 ms
> .
> [INFO]: Torch Forward pass peak memory: 26718.27 MB
> [INFO]: Torch Backward pass peak memory: 57095.80 MB
> .
> [INFO]: Kernel Forward pass peak memory: 1376.98 MB
> [INFO]: Kernel Backward pass peak memory: 2984.55 MB
> .                                                                                                                                                                                                                [INFO]: Forward pass: Torch implementation average time: 52.30 ms
> [INFO]: Backward pass: torch implementation average time: 139.77 ms
> [INFO]: Forward pass: Kernel implementation average time: 520.16 ms
> [INFO]: Backward pass: kernel implementation average time: 554.70 ms
> .
> [INFO]: Torch Forward pass peak memory: 19123.71 MB
> [INFO]: Torch Backward pass peak memory: 26718.09 MB
> .
> [INFO]: Kernel Forward pass peak memory: 1369.42 MB
> [INFO]: Kernel Backward pass peak memory: 2984.37 MB
> ```
> 
> The tests were run using the [Nvidia NeMo docker container version 25.07](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo?version=25.07.gpt_oss) . Please find attached the list of environment variables. [collect_env.txt](https://github.com/user-attachments/files/22401195/collect_env.txt)
> 
> It seems that the kernel implementation time is significantly larger than the corresponding time for PyTorch, unlike the screenshot in the earlier message. @Jianbing-D , could you please clarify the hardware and environment used for your Pytest runs?

Hi @nagarajankarthik ,

Those numbers were collected with `nvcr.io/nvidia/pytorch:25.04-py3` and `NVIDIA H100 GPU`. You may follow these steps to reproduce them.
```shell
# start a container with aforementioned image
# clone this feature branch
$ pip uninstall transformer_engine 
$ export PYTHONUSERBASE=/tmp/python
# replace 100a with the compute capability of your target GPU
$ unset PIP_CONSTRAINT && NVTE_CUDA_ARCHS="100a" NVTE_BUILD_THREADS_PER_JOB=8 NVTE_FRAMEWORK=pytorch pip install  --no-build-isolation -v -e ./TransformerEngine
$ python3 -m pytest -v -s $PWD/TransformerEngine/tests/pytorch/test_linear_cross_entropy.py
```
Logs should be like
```text
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropyWithTokenEntropy::test_correctness
[INFO]: Forward pass: Torch implementation average time: 94.08 ms
[INFO]: Backward pass: torch implementation average time: 197.11 ms
[INFO]: Forward pass: Kernel implementation average time: 58.96 ms
[INFO]: Backward pass: kernel implementation average time: 85.47 ms
PASSED
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropyWithTokenEntropy::test_torch_storage
[INFO]: Torch Forward pass peak memory: 26718.27 MB
[INFO]: Torch Backward pass peak memory: 57095.80 MB
PASSED
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropyWithTokenEntropy::test_kernel_storage
[INFO]: Kernel Forward pass peak memory: 1376.98 MB
[INFO]: Kernel Backward pass peak memory: 2984.55 MB
PASSED
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropy::test_correctness
[INFO]: Forward pass: Torch implementation average time: 56.69 ms
[INFO]: Backward pass: torch implementation average time: 125.68 ms
[INFO]: Forward pass: Kernel implementation average time: 60.12 ms
[INFO]: Backward pass: kernel implementation average time: 83.06 ms
PASSED
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropy::test_torch_storage
[INFO]: Torch Forward pass peak memory: 19123.71 MB
[INFO]: Torch Backward pass peak memory: 26718.09 MB
PASSED
TransformerEngine/tests/pytorch/test_linear_cross_entropy.py::TestLinearCrossEntropy::test_kernel_storage
[INFO]: Kernel Forward pass peak memory: 1369.42 MB
[INFO]: Kernel Backward pass peak memory: 2984.37 MB
PASSED
```

However, based on your experiment results, it seems like newer version would impact those numbers.

I will do further research about this.
